lib/cadet/accounts/notification.ex

```diff
@@ -189,4 +193,116 @@ defmodule Cadet.Accounts.Notification do
       assessment_id: assessment_id
     })
   end
+
+  @doc """
+  Writes a notification that a student's submission has been
+  autograded successfully. (for the student)
+  """
+  @spec write_notification_when_autograded(integer() | String.t()) :: Ecto.Changeset.t()
+  def write_notification_when_autograded(submission_id) do
+    submission =
+      Submission
+      |> Repo.get_by(id: submission_id)
+
+    params = %{
+      type: :autograded,
+      read: false,
+      role: :student,
+      user_id: submission.student_id,
+      assessment_id: submission.assessment_id,
+      submission_id: submission_id
+    }
+
+    write(params)
+  end
+
+  @doc """
+  Writes a notification that a student's submission has been
+  manually graded successfully by an Avenger or other teaching staff.
+  (for the student)
+  """
+  @spec write_notification_when_manually_graded(integer() | String.t()) :: Ecto.Changeset.t()
+  def write_notification_when_manually_graded(submission_id) do
+    submission =
+      Submission
+      |> Repo.get_by(id: submission_id)
+
+    params = %{
+      type: :graded,
+      read: false,
+      role: :student,
+      user_id: submission.student_id,
+      assessment_id: submission.assessment_id,
+      submission_id: submission_id
+    }
+
+    write(params)
+  end
+
+  @doc """
+  Writes a notification to all students that a new assessment is available.
+  """
+  @spec write_notification_for_new_assessment(integer() | String.t()) ::
+          {:ok, any()}
+          | {:error, any()}
+          | {:error, Ecto.Multi.name(), any(), %{required(Ecto.Multi.name()) => any()}}
+  def write_notification_for_new_assessment(assessment_id) do
+    assessment =
+      Assessment
+      |> Repo.get_by(id: assessment_id)
+
+    notification_multi = Multi.new()
+
+    if Cadet.Assessments.is_open?(assessment) do
+      User
+      |> where(role: ^:student)
+      |> Repo.all()
+      |> Enum.each(fn %User{id: student_id} ->
+        params = %{
+          type: :new,
+          read: false,
+          role: :student,
+          user_id: student_id,
+          assessment_id: assessment_id
+        }
+
+        changes =
+          %Notification{}
+          |> changeset(params)
+
+        Multi.insert(
+          notification_multi,
+          String.to_atom("notify_new_for_student_#{student_id}"),
+          changes
+        )
```

Comment:
Good try, but the multi here isn't saved as a variable anywhere. In the end the multi that get transacted is an empty multi...